diff --git a/agreement/actions.go b/agreement/actions.go
index f2b85d45..66913595 100644
--- a/agreement/actions.go
+++ b/agreement/actions.go
@@ -547,6 +547,7 @@ func (c checkpointAction) do(ctx context.Context, s *Service) {
 	}
 	if c.Err == nil {
 		s.log.with(logEvent).Infof("checkpoint at (%v, %v, %v)", c.Round, c.Period, c.Step)
+		SpecTracePersistState(c.Round, c.Period, c.Step)
 	} else {
 		s.log.with(logEvent).Errorf("checkpoint at (%v, %v, %v) failed : %v", c.Round, c.Period, c.Step, c.Err)
 		if c.done != nil {
diff --git a/agreement/player.go b/agreement/player.go
index 1f6fe71a..a722753e 100644
--- a/agreement/player.go
+++ b/agreement/player.go
@@ -160,6 +160,7 @@ func (p *player) handleFastTimeout(r routerHandle, e timeoutEvent) []action {
 	if p.FastRecoveryDeadline == 0 {
 		// don't vote the first time
 		p.FastRecoveryDeadline = lower + delta + lambda // add lambda for extra delay the first time
+		SpecTraceHandleFastTimeoutPrimer(p)
 		return nil
 	}
 	p.FastRecoveryDeadline = lower + delta
@@ -185,30 +186,37 @@ func (p *player) issueSoftVote(r routerHandle, deadlineTimeout time.Duration) (a
 		// If we arrive due to fast-forward/soft threshold; then answer.Bottom = false and answer.Proposal = bottom
 		// and we should soft-vote normally (not based on the starting value)
 		a.Proposal = nextStatus.Proposal
+		SpecTraceIssueSoftVote(p, &a)
 		return append(actions, a)
 	}
 
 	if a.Proposal == bottom {
 		// did not see anything: do not vote
+		SpecTraceIssueSoftVote(p, nil)
 		return nil
 	}
 
 	if p.Period > a.Proposal.OriginalPeriod {
 		// leader sent reproposal: vote if we saw a quorum for that hash, even if we saw nextStatus.Bottom
 		if nextStatus.Proposal != bottom && nextStatus.Proposal == a.Proposal {
+			SpecTraceIssueSoftVote(p, &a)
 			return append(actions, a)
 		}
+		SpecTraceIssueSoftVote(p, nil)
 		return nil
 	}
 
 	// original proposal: vote for it
+	SpecTraceIssueSoftVote(p, &a)
 	return append(actions, a)
 }
 
 // A committableEvent is the trigger for issuing a cert vote.
 func (p *player) issueCertVote(r routerHandle, e committableEvent) action {
 	r.t.timeR().RecStep(p.Period, cert, e.Proposal)
-	return pseudonodeAction{T: attest, Round: p.Round, Period: p.Period, Step: cert, Proposal: e.Proposal}
+	pa := pseudonodeAction{T: attest, Round: p.Round, Period: p.Period, Step: cert, Proposal: e.Proposal}
+	SpecTraceIssueCertVote(p, pa)
+	return pa
 }
 
 func (p *player) issueNextVote(r routerHandle, deadlineTimeout time.Duration) []action {
@@ -238,6 +246,7 @@ func (p *player) issueNextVote(r routerHandle, deadlineTimeout time.Duration) []
 	_, upper := p.Step.nextVoteRanges(deadlineTimeout)
 	p.Napping = false
 	p.Deadline = Deadline{Duration: upper, Type: TimeoutDeadline}
+	SpecTraceIssueNextVote(p, a)
 	return actions
 }
 
@@ -270,6 +279,7 @@ func (p *player) issueFastVote(r routerHandle) (actions []action) {
 		a.Step = down
 	}
 
+	SpecTraceIssueFastVote(p, a)
 	return append(actions, a)
 }
 
@@ -311,6 +321,7 @@ func (p *player) updateCredentialArrivalHistory(r routerHandle, ver protocol.Con
 	}
 
 	p.lowestCredentialArrivals.store(re.LowestIncludingLate.validatedAt)
+	SpecTraceRecordCredentialArrival(p)
 	return re.LowestIncludingLate.validatedAt
 }
 
@@ -320,11 +331,13 @@ func (p *player) calculateFilterTimeout(ver protocol.ConsensusVersion, tracer *t
 	if dynamicFilterCredentialArrivalHistory <= 0 || p.Period != 0 {
 		// Either dynamic filter timeout is disabled, or we're not in period 0
 		// and therefore, can't use dynamic timeout
+		SpecTraceCalculateFilterTimeoutDefault(p)
 		return FilterTimeout(p.Period, ver)
 	}
 	defaultTimeout := FilterTimeout(0, ver)
 	if !p.lowestCredentialArrivals.isFull() {
 		// not enough samples, use the default
+		SpecTraceCalculateFilterTimeoutDefault(p)
 		return defaultTimeout
 	}
 
@@ -340,9 +353,11 @@ func (p *player) calculateFilterTimeout(ver protocol.ConsensusVersion, tracer *t
 		// If the dynamic filter timeout is disabled, return the default filter
 		// timeout (after logging what the timeout would have been, if this
 		// feature were enabled).
+		SpecTraceCalculateFilterTimeoutDefault(p)
 		return defaultTimeout
 	}
 
+	SpecTraceCalculateFilterTimeoutShort(p)
 	return clampedTimeout
 }
 
@@ -363,6 +378,7 @@ func (p *player) handleThresholdEvent(r routerHandle, e thresholdEvent) []action
 			a0.voteValidatedAt = p.updateCredentialArrivalHistory(r, e.Proto)
 			a0.dynamicFilterTimeout = p.dynamicFilterTimeout
 			actions = append(actions, a0)
+			SpecTraceHandleCertThresholdLocal(p, e)
 			as := p.enterRound(r, e, p.Round+1)
 			return append(actions, as...)
 		}
@@ -385,6 +401,7 @@ func (p *player) handleThresholdEvent(r routerHandle, e thresholdEvent) []action
 		}
 		ec := r.dispatch(*p, e, proposalMachine, p.Round, p.Period, 0)
 		if ec.t() == proposalCommittable && p.Step <= cert {
+			SpecTraceHandleSoftThresholdSamePeriod(p, e)
 			actions = append(actions, p.issueCertVote(r, ec.(committableEvent)))
 		}
 		return actions
@@ -421,6 +438,7 @@ func (p *player) enterPeriod(r routerHandle, source thresholdEvent, target perio
 		// calculation mechanism.
 		p.lowestCredentialArrivals.reset()
 	}
+	SpecTraceEnterPeriod(p, source)
 	p.Deadline = Deadline{Duration: p.calculateFilterTimeout(source.Proto, r.t), Type: TimeoutFilter}
 
 	// update tracer state to match player
@@ -467,6 +485,8 @@ func (p *player) enterRound(r routerHandle, source event, target round) []action
 	p.Napping = false
 	p.FastRecoveryDeadline = 0 // set immediately
 
+	SpecTraceEnterRound(p)
+
 	switch source := source.(type) {
 	case roundInterruptionEvent:
 		p.Deadline = Deadline{Duration: p.calculateFilterTimeout(source.Proto.Version, r.t), Type: TimeoutFilter}
@@ -522,6 +542,7 @@ func (p *player) partitionPolicy(r routerHandle) (actions []action) {
 		r.t.logBundleBroadcast(*p, b)
 		a0 := broadcastBundleAction(b)
 		actions = append(actions, a0)
+		SpecTracePartitionPolicyRebroadcast(p, b)
 	}
 
 	// On resynchronization, first try relaying the staged proposal from the same period as
diff --git a/agreement/proposalManager.go b/agreement/proposalManager.go
index 7230ff5a..bd43919f 100644
--- a/agreement/proposalManager.go
+++ b/agreement/proposalManager.go
@@ -172,6 +172,8 @@ func (m *proposalManager) handleMessageEvent(r routerHandle, p player, e filtera
 			r.t.timeRPlus1().RecVoteReceived(v)
 		}
 
+		SpecTraceReceiveProposal(&p, v)
+
 		e := r.dispatch(p, e.messageEvent, proposalMachineRound, v.R.Round, v.R.Period, 0)
 
 		if keepForLateCredentialTracking {
diff --git a/agreement/proposalTracker.go b/agreement/proposalTracker.go
index f5ca4ba5..bf9e4203 100644
--- a/agreement/proposalTracker.go
+++ b/agreement/proposalTracker.go
@@ -202,7 +202,11 @@ func (t *proposalTracker) handle(r routerHandle, p player, e event) event {
 
 	case softThreshold, certThreshold:
 		e := e.(thresholdEvent)
+		prevStaging := t.Staging
 		t.Staging = e.Proposal
+		if prevStaging == bottom && e.Proposal != bottom {
+			SpecTraceUpdateStaging(&p, e)
+		}
 
 		return proposalAcceptedEvent{
 			Round:    e.Round,
diff --git a/agreement/pseudonode.go b/agreement/pseudonode.go
index 75ec5efc..d050258b 100644
--- a/agreement/pseudonode.go
+++ b/agreement/pseudonode.go
@@ -479,6 +479,7 @@ func (t pseudonodeVotesTask) execute(verifier *AsyncVoteVerifier, quit chan stru
 	// push results into channel.
 verifiedVotesLoop:
 	for _, r := range verifiedResults {
+		SpecTraceBroadcastVote(r.v)
 		for {
 			select {
 			case t.out <- messageEvent{T: voteVerified, Input: r.message, Err: makeSerErr(r.err)}:
@@ -585,6 +586,7 @@ func (t pseudonodeProposalsTask) execute(verifier *AsyncVoteVerifier, quit chan
 	// push results into channel.
 verifiedVotesLoop:
 	for _, r := range verifiedVotes {
+		SpecTraceProposeBlock(r.v)
 		for {
 			select {
 			case t.out <- messageEvent{T: voteVerified, Input: r.message, Err: makeSerErr(r.err)}:
diff --git a/agreement/voteAggregator.go b/agreement/voteAggregator.go
index c1e73d72..5eaa03f5 100644
--- a/agreement/voteAggregator.go
+++ b/agreement/voteAggregator.go
@@ -115,6 +115,8 @@ func (agg *voteAggregator) handle(r routerHandle, pr player, em event) (res even
 			r.t.timeRPlus1().RecVoteReceived(v)
 		}
 
+		SpecTraceReceiveVote(&pr, v)
+
 		deliver := voteAcceptedEvent{Vote: v, Proto: e.Proto.Version}
 		tE := r.dispatch(pr, deliver, voteMachineRound, v.R.Round, v.R.Period, v.R.Step)
 		if tE.t() == none {
diff --git a/agreement/voteAuxiliary.go b/agreement/voteAuxiliary.go
index 64a16a1e..9a043505 100644
--- a/agreement/voteAuxiliary.go
+++ b/agreement/voteAuxiliary.go
@@ -70,10 +70,17 @@ func (t *voteTrackerPeriod) handle(r routerHandle, p player, e event) event {
 		return e
 	case nextThreshold:
 		// cache next thresholds in response
-		if e.(thresholdEvent).Proposal == bottom {
-			t.Cached.Bottom = true
+		te := e.(thresholdEvent)
+		if te.Proposal == bottom {
+			if !t.Cached.Bottom {
+				t.Cached.Bottom = true
+				SpecTraceUpdateNextThresholdCache(&p, te)
+			}
 		} else {
-			t.Cached.Proposal = e.(thresholdEvent).Proposal
+			if t.Cached.Proposal != te.Proposal {
+				t.Cached.Proposal = te.Proposal
+				SpecTraceUpdateNextThresholdCache(&p, te)
+			}
 		}
 		return emptyEvent{}
 	case nextThresholdStatusRequest:
@@ -145,6 +152,7 @@ func (t *voteTrackerRound) handle(r routerHandle, p player, e event) event {
 		if e.(thresholdEvent).fresherThan(t.Freshest) {
 			t.Freshest = e.(thresholdEvent)
 			t.Ok = true
+			SpecTraceUpdateFreshest(&p, e.(thresholdEvent))
 			return e
 		}
 		return emptyEvent{}
