diff --git a/lib/ring/rte_ring_elem_pvt.h b/lib/ring/rte_ring_elem_pvt.h
index 6eafae121f..0d5e9fdcea 100644
--- a/lib/ring/rte_ring_elem_pvt.h
+++ b/lib/ring/rte_ring_elem_pvt.h
@@ -10,6 +10,10 @@
 #ifndef _RTE_RING_ELEM_PVT_H_
 #define _RTE_RING_ELEM_PVT_H_
 
+#ifdef DPDK_TLA_TRACE
+#include "rte_ring_tla_trace.h"
+#endif
+
 #if defined(RTE_TOOLCHAIN_GCC) && (GCC_VERSION >= 120000)
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wstringop-overflow"
@@ -416,9 +420,21 @@ __rte_ring_do_enqueue_elem(struct rte_ring *r, const void *obj_table,
 	if (n == 0)
 		goto end;
 
+#ifdef DPDK_TLA_TRACE
+	__tla_trace_reserve_prod(r, n);
+#endif
+
 	__rte_ring_enqueue_elems(r, prod_head, obj_table, esize, n);
 
+#ifdef DPDK_TLA_TRACE
+	__tla_trace_write_data(r);
+#endif
+
 	__rte_ring_update_tail(&r->prod, prod_head, prod_next, is_sp, 1);
+
+#ifdef DPDK_TLA_TRACE
+	__tla_trace_publish_tail(r);
+#endif
 end:
 	if (free_space != NULL)
 		*free_space = free_entries - n;
@@ -463,10 +479,22 @@ __rte_ring_do_dequeue_elem(struct rte_ring *r, void *obj_table,
 	if (n == 0)
 		goto end;
 
+#ifdef DPDK_TLA_TRACE
+	__tla_trace_reserve_cons(r, n);
+#endif
+
 	__rte_ring_dequeue_elems(r, cons_head, obj_table, esize, n);
 
+#ifdef DPDK_TLA_TRACE
+	__tla_trace_write_data(r);
+#endif
+
 	__rte_ring_update_tail(&r->cons, cons_head, cons_next, is_sc, 0);
 
+#ifdef DPDK_TLA_TRACE
+	__tla_trace_publish_tail(r);
+#endif
+
 end:
 	if (available != NULL)
 		*available = entries - n;
diff --git a/lib/ring/rte_ring_hts_elem_pvt.h b/lib/ring/rte_ring_hts_elem_pvt.h
index a01089d15d..2215aa33df 100644
--- a/lib/ring/rte_ring_hts_elem_pvt.h
+++ b/lib/ring/rte_ring_hts_elem_pvt.h
@@ -12,6 +12,10 @@
 
 #include <rte_stdatomic.h>
 
+#ifdef DPDK_TLA_TRACE
+#include "rte_ring_tla_trace.h"
+#endif
+
 /**
  * @file rte_ring_hts_elem_pvt.h
  * It is not recommended to include this file directly,
@@ -216,8 +220,17 @@ __rte_ring_do_hts_enqueue_elem(struct rte_ring *r, const void *obj_table,
 	n =  __rte_ring_hts_move_prod_head(r, n, behavior, &head, &free);
 
 	if (n != 0) {
+#ifdef DPDK_TLA_TRACE
+		__tla_trace_reserve_prod(r, n);
+#endif
 		__rte_ring_enqueue_elems(r, head, obj_table, esize, n);
+#ifdef DPDK_TLA_TRACE
+		__tla_trace_write_data(r);
+#endif
 		__rte_ring_hts_update_tail(&r->hts_prod, head, n, 1);
+#ifdef DPDK_TLA_TRACE
+		__tla_trace_publish_tail(r);
+#endif
 	}
 
 	if (free_space != NULL)
@@ -257,8 +270,17 @@ __rte_ring_do_hts_dequeue_elem(struct rte_ring *r, void *obj_table,
 	n = __rte_ring_hts_move_cons_head(r, n, behavior, &head, &entries);
 
 	if (n != 0) {
+#ifdef DPDK_TLA_TRACE
+		__tla_trace_reserve_cons(r, n);
+#endif
 		__rte_ring_dequeue_elems(r, head, obj_table, esize, n);
+#ifdef DPDK_TLA_TRACE
+		__tla_trace_write_data(r);
+#endif
 		__rte_ring_hts_update_tail(&r->hts_cons, head, n, 0);
+#ifdef DPDK_TLA_TRACE
+		__tla_trace_publish_tail(r);
+#endif
 	}
 
 	if (available != NULL)
diff --git a/lib/ring/rte_ring_rts_elem_pvt.h b/lib/ring/rte_ring_rts_elem_pvt.h
index 0280369b21..bf3ce7f3b6 100644
--- a/lib/ring/rte_ring_rts_elem_pvt.h
+++ b/lib/ring/rte_ring_rts_elem_pvt.h
@@ -10,6 +10,10 @@
 #ifndef _RTE_RING_RTS_ELEM_PVT_H_
 #define _RTE_RING_RTS_ELEM_PVT_H_
 
+#ifdef DPDK_TLA_TRACE
+#include "rte_ring_tla_trace.h"
+#endif
+
 /**
  * @file rte_ring_rts_elem_pvt.h
  * It is not recommended to include this file directly,
@@ -231,8 +235,17 @@ __rte_ring_do_rts_enqueue_elem(struct rte_ring *r, const void *obj_table,
 	n =  __rte_ring_rts_move_prod_head(r, n, behavior, &head, &free);
 
 	if (n != 0) {
+#ifdef DPDK_TLA_TRACE
+		__tla_trace_reserve_prod(r, n);
+#endif
 		__rte_ring_enqueue_elems(r, head, obj_table, esize, n);
+#ifdef DPDK_TLA_TRACE
+		__tla_trace_write_data(r);
+#endif
 		__rte_ring_rts_update_tail(&r->rts_prod);
+#ifdef DPDK_TLA_TRACE
+		__tla_trace_publish_tail(r);
+#endif
 	}
 
 	if (free_space != NULL)
@@ -272,8 +285,17 @@ __rte_ring_do_rts_dequeue_elem(struct rte_ring *r, void *obj_table,
 	n = __rte_ring_rts_move_cons_head(r, n, behavior, &head, &entries);
 
 	if (n != 0) {
+#ifdef DPDK_TLA_TRACE
+		__tla_trace_reserve_cons(r, n);
+#endif
 		__rte_ring_dequeue_elems(r, head, obj_table, esize, n);
+#ifdef DPDK_TLA_TRACE
+		__tla_trace_write_data(r);
+#endif
 		__rte_ring_rts_update_tail(&r->rts_cons);
+#ifdef DPDK_TLA_TRACE
+		__tla_trace_publish_tail(r);
+#endif
 	}
 
 	if (available != NULL)
